run/radius2john.py

```diff
@@ -1,68 +1,163 @@
-#!/usr/bin/env python
+#!/usr/bin/env python3
 
-# This software is Copyright (c) 2019 Maxime GOYETTE <maxgoyette0-at-gmail.com>,
+# This software is Copyright (c) 2024, k4amos <k4amos at proton.me>
 # and it is hereby released to the general public under the following terms:
+#
 # Redistribution and use in source and binary forms, with or without
 # modification, are permitted.
-#
+
+# This script is essentially a python version of radius2john.pl written by Didier ARENZANA. 
+# The previous version of radius2john.py was written by Maxime GOYETTE <maxgoyette0-at-gmail.com>
+
+# ---
+
 # Utility to bruteforce RADIUS shared-secret
 # Usage: ./radius2john.py <pcap files>
 #
 # This script depends on Scapy (https://scapy.net)
 # To install: pip install --user scapy
 
+# ---
+
+# Application of two  methods described in http://www.untruth.org/~josh/security/radius/radius-auth.html :
+# "3.3 User-Password Attribute Based Shared Secret Attack"
+# "3.1 Response Authenticator Based Shared Secret Attack"
+
+# For attack 3.3 :
+# We try authentications using a known password, and sniff the radius packets to a pcap file.
+# This script reads access-request in the pcap file, and dumps the md5(RA+secret) and RA, in a john-friendly format.
+
+# For attack 3.1:
+# We don't need to try authentications. Just sniff the radius packets in a pcap file.
+# This script reads the pcap file, matches radius responses with the corresponding all_requests,
+# and dumps md5 and salt as needed.
+
+import binascii
+import sys
+import argparse
+
 try:
-    from scapy.all import *
+    import scapy.all as scapy
 except ImportError:
-    print("scapy is missing, run 'pip install --user scapy' to install it!")
-    exit(1)
+    print(
+        "Scapy seems to be missing, run 'pip install --user scapy' to install it"
+    )
+    sys.exit(1)
+
+def read_file(args, filename):
+    packets = scapy.rdpcap(filename)
+    for packet in packets:
+        process_packet(args, packet)
+
+
+def process_packet(args, packet):
+    if packet.haslayer(scapy.IP) and packet.haslayer(scapy.UDP):
+        ip_layer = packet[scapy.IP]
+        udp_layer = packet[scapy.UDP]
+
+        if udp_layer.dport in [1812, 1813] or udp_layer.sport in [1812, 1813]: 
+            process_radius(args, ip_layer, bytes(udp_layer.payload))
+
+
+def process_radius(args, ip, udpdata):
+    radius_packet = scapy.Radius(udpdata)
+
+    if radius_packet.code in [1, 4]:  # Access-Request, Accounting-Request
+
+        if args['password'] is not None:
+
+            user_name, user_hash = None, None
+
+            for attr in radius_packet.attributes:
+                if attr.name == "User-Name":
+                    user_name = attr.value.decode("utf-8")
+
+                if attr.name == "User-Password":
+                    user_hash = attr.value
+
+            if user_hash is not None:
+                if (args['login'] is not None and user_name == args['login']) or args['login'] is None:
```

Comment:
I think this can be simplified to `if args['login'] is None or user_name == args['login']`